File: slider.js

function Slider(instrument, container, key, min, max, step) {
    this.key = key;
    this.min = min;
    this.max = max;
    this.step = step;
    this.instrument = instrument;
    this.container = $('<div class="slider"/>');
    this.label = $('<div class="label"/>');
    this.range = $('<div class="range"/>');
    this.handle = $('<div class="handle"/>');
    this.position = 0;

    this.label.html(this.key);
    this.container.append(this.label).append(this.range);
    this.range.append(this.handle);
    container.append(this.container);

    this.width = this.range.width() - this.handle.width();

    this.automation = new Automation(this.instrument, container, key);

    this.label.click(this.onClickLabel.bind(this));

    this.tracker = new TouchTracker(this, this.range.get(0), this.handleEvent.bind(this));
}

Slider.prototype.onClickLabel = function(event) {
    this.automation.toggle();
    event.preventDefault();
};

Slider.prototype.setValue = function(value) {
    value = Math.max(this.min, Math.min(this.max, value));
    this.value = value;
    this.position = this.width * ((value - this.min) / (this.max - this.min));
    this.handle.css('marginLeft', this.position + 'px');     
};

Slider.prototype.handleEvent = function(event) {
    var sliderLeft = this.range.offset().left;
    var left = event.pageX - sliderLeft - this.handle.width() / 2;
    var value = this.min + (left / this.width) * (this.max - this.min);

    value = Math.floor(value / this.step) * this.step;

    if (value != this.value) {
        this.setValue(value);

        controller.send('parameter', this.instrument.index, this.key, this.value);
    }
};

File: slider.rb
class Slider
  attr_reader :key, :value, :min, :max, :step, :automation

  def initialize(instrument, key, value, min, max, step)
    @instrument = instrument
    @automation = Array.new(8) { Array.new(16) { 0 } }
    @key = key
    @value = value
    @min = min
    @max = max
    @step = step
  end

  def messages
    messages = []
    messages << constructor_message
    messages << parameter_message

    @automation.each_with_index do |pattern, clip|
      pattern.each_with_index do |value, index|
        messages << automation_message(clip, index, value)
      end
    end

    messages
  end

  def constructor_message
    Message.new("/slider", "isfff", @instrument.index, @key, @min, @max, @step)
  end

  def parameter_message
    Message.new("/parameter", "isf", @instrument.index, @key, @value)
  end

  def automation_message(clip, index, value)
    Message.new("/automation", "isiif", @instrument.index, @key, clip, index, value)
  end

  def set_value(value)
    @value = value.to_f
    $sender.send(parameter_message)
  end

  def automation(clip, index, value)
    @automation[clip.to_i][index.to_i] = value.to_f
    $sender.send(automation_message(clip, index, value))
  end
  
end
